src/apps/mixed_fem_helper.hpp

```diff
@@ -0,0 +1,942 @@
+    tbasis_y[m] = tyx*basis_x[m] + tyy*basis_y[m] + tyz*basis_z[m];
+    tbasis_z[m] = tzx*basis_x[m] + tzy*basis_y[m] + tzz*basis_z[m];
+  }
+}
+
+template<int M, int P>
+RAJA_HOST_DEVICE
+constexpr void inner_product(
+  const double weight,
+  const double (&basis_1_x)[M],
+  const double (&basis_1_y)[M],
+  const double (&basis_1_z)[M],
+  const double (&basis_2_x)[P],
+  const double (&basis_2_y)[P],
+  const double (&basis_2_z)[P],
+  double (&matrix)[P][M],
+  const bool is_symmetric)
+{
+  // inner product is <basis_2, basis_1>
+  for (int p = 0; p < P; p++) {
+
+    const double txi = basis_2_x[p];
+    const double tyi = basis_2_y[p];
+    const double tzi = basis_2_z[p];
+
+    const int m0 = (is_symmetric && (M == P)) ? p : 0;
+
+    for (int m = m0; m < M; m++) {
+
+      const double txj = basis_1_x[m];
+      const double tyj = basis_1_y[m];
+      const double tzj = basis_1_z[m];
+
+      matrix[p][m] += weight*(txi*txj + tyi*tyj + tzi*tzj);
+
+      if(is_symmetric && (M == P) && (m > m0))
+      {
+        matrix[m][p] = matrix[p][m];
+      }
+    }
+  }
+}
+
+RAJA_HOST_DEVICE
+inline void bad_zone_algorithm(
+  const double detj_unfixed,
+  const double detj_cc,
+  const double detj_tol,
+  double& detj,
+  double& abs_detj,
+  double& inv_detj)
+{
+  detj = (fabs( detj_unfixed/detj_cc ) < detj_tol) ?
+                               detj_cc : detj_unfixed ;
+  abs_detj = fabs(detj);
+
+  // Note that this uses a potentially negative detj
+
+  inv_detj = 1.0/(detj + ptiny);
+}
+
+RAJA_HOST_DEVICE
+inline void jacobian_inv(
+  const double jxx,
+  const double jxy,
+  const double jxz,
+  const double jyx,
+  const double jyy,
+  const double jyz,
+  const double jzx,
+  const double jzy,
+  const double jzz,
+  const double detj_cc,
+  const double detj_tol,
+  double &jinvxx,
+  double &jinvxy,
+  double &jinvxz,
+  double &jinvyx,
+  double &jinvyy,
+  double &jinvyz,
+  double &jinvzx,
+  double &jinvzy,
+  double &jinvzz,
+  double &detj_unfixed,
+  double &detj,
+  double &abs_detj,
+  double &inv_detj)
+{
+  // Compute determinant of Jacobian matrix at this quadrature point
+  detj_unfixed = compute_detj(jxx, jxy, jxz,
+                             jyx, jyy, jyz,
+                             jzx, jzy, jzz);
+
+  bad_zone_algorithm(detj_unfixed, detj_cc, detj_tol, detj, abs_detj, inv_detj);
+
+  jinvxx =  (jyy*jzz - jyz*jzy)*inv_detj;
+  jinvxy =  (jxz*jzy - jxy*jzz)*inv_detj;
+  jinvxz =  (jxy*jyz - jxz*jyy)*inv_detj;
+
+  jinvyx =  (jyz*jzx - jyx*jzz)*inv_detj;
+  jinvyy =  (jxx*jzz - jxz*jzx)*inv_detj;
+  jinvyz =  (jxz*jyx - jxx*jyz)*inv_detj;
+
+  jinvzx =  (jyx*jzy - jyy*jzx)*inv_detj;
+  jinvzy =  (jxy*jzx - jxx*jzy)*inv_detj;
+  jinvzz =  (jxx*jyy - jxy*jyx)*inv_detj;
+}
+
+RAJA_HOST_DEVICE
+constexpr double Jzx(
+  const double  (&x)[NB],
+  const double  (&y)[NB],
+  const double  (&z)[NB],
+  const double  tmpxy,
+  const double  xloctmpy,
+  const double  xyloc,
+  const double  tmpxyloc)
+{
+  (void) x[0];
```

Comment:
1. Is the cast to void just for compiler warning? 
2. Can we remove the y and z from this scope block?
3. Is the bottom section of this scope block unreachable?

Reply:
1.  Yes
3.  The return statement is the last line that is called.

Reply:
Done